hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,740 @@
+  if (Delta==0)
+  {
+    Delta= new CircularBuffer(INTEGRAL_DURATION);
+    DeltaTemps[index] = Delta;
+  }
+  //put last value
+  Delta->Put(DeltaTemp);
+  //need heating
+  if (DeltaTemp>0)
+  {
+    //PID regulation
+    // coef integral is the sum of delta temperature since last 10 minutes.
+	  PowerModulation = int(DeltaTemp * CoefProportional + Delta->GetSum() * CoefIntegral /INTEGRAL_DURATION );
+  }
+	else
+		Delta->Clear(); //clear integral part
+
+  if (PowerModulation>100) PowerModulation=100;
+  if (PowerModulation<0  ) PowerModulation=0;
+  return PowerModulation;
+}                    
+
+//return the output command & level from cmd value
+//cmd ='On;Off;Set Level XX "
+void getCommand(std::string &Cmd, std::string &OutCmd, int &level)
+{
+	OutCmd = Cmd; level = 0;
+
+	if (Cmd == "Off"){
+		level = 0;
+	}
+	else if (Cmd == "On") {
+		level = 100;
+	}
+	else {
+		std::vector<std::string> results;
+		StringSplit(Cmd, " ", results);
+        if (results.size()==3){
+		    if ( (results[0] == "Set") && (results[1] == "Level")) {
+			    OutCmd = "Set Level"; 
+                try{
+                    level = std::stoi (results[2] ) ;
+                }
+                catch(...){
+                    _log.Log(LOG_ERROR,"Invalid Switch command : %s ", Cmd.c_str() );
+                }
+            }
+            else
+ 				_log.Log(LOG_ERROR,"Invalid Switch command : %s ", Cmd.c_str() );
+		}
+        else
+ 		    _log.Log(LOG_ERROR,"Invalid Switch command : %s ", Cmd.c_str() );
+	}
+}
+
+//get level from imput cmd : selectof value 
+int translateCmd(std::string &Cmd,TOptionMap & options)
+{
+int Level=-1;
+	Level = GetSelectorSwitchLevel(options, Cmd);
+	if (Level >= 0) { 
+        Cmd = "Set Level " + std::to_string(Level);
+    }
+    return Level;
+}
+void VirtualThermostat::ScheduleThermostat(int Minute )
+{
+
+    int SwitchValue = 0;
+    int nValue, lastSwitchValue;
+    std::string RoomTemperatureName;
+    double ThermostatSetPoint = 0;
+    const char* ThermostatName;
+    struct tm LastUpdateTime;
+    std::string sValue, Options;
+    float RoomTemperature = 0;
+    int ThermostatId;
+    int PowerModulation = 0;
+    int lastPowerModulation;
+    double lastTemp;
+
+    long  SwitchIdx;
+    std::string SwitchIdxStr;
+
+    double CoefProportional, CoefIntegral;
+    std::string TemperatureId;
+    std::string OnCmd;
+    std::string OffCmd;
+    long DeviceID;
+    std::string sDeviceID;
+
+    int switchtype;
+
+try
+{	
+//select all the Virtuan device
+  TSqlQueryResult result=m_sql.safe_query("SELECT A.Name,A.ID,A.Type,A.SubType,A.nValue,A.sValue, A.Options, A.HardwareID , B.Type,A.DeviceID     FROM DeviceStatus as A LEFT OUTER JOIN Hardware as B ON (B.ID==A.HardwareID) where (B.type == %d )",HTYPE_VirtualThermostat );
+
+	//for all the thermostat switch
+    unsigned int nbDevices     =  result.size() ;
+    double scrollDevice  =  double(MODULATION_DURATION)/nbDevices;
+
+	for (unsigned int i=0;i<nbDevices;i++)
```

Comment:
better to have nbDevices of type size_t

you can also use

```
for (const auto &itt : result)
{
}

```

no need to check if result is empty